### PR TITLE
Packet: improve bootstrap iptables rules

### PR DESCRIPTION
**Pull request by invidian**
This PR further locks down bootstrap iptables rules. The disadvantage though, which we need to consider is, that if user updates some of those CIDRs and rely on them, then Calico rules are flushed for whatever reason, the original rules will be applied (from node creation time), unless the node is replaced during the process. We need to make sure, that this is the behavior we want.

Alternatively, changing CIDRs would require node upgrade rollout we plan to implement. 

Refs #137 #8

### CODE REVIEW

File: assets/lokomotive-kubernetes/packet/flatcar-linux/kubernetes/cl/controller.yaml.tmpl [hunk 1]
```diff
@@ -266,11 +266,13 @@ storage:
           :FORWARD ACCEPT [0:0]
           :OUTPUT ACCEPT [0:0]
           -A INPUT -i lo -j ACCEPT
-          -A INPUT -p tcp --dport 22 -j ACCEPT
-          -A INPUT -p tcp --dport 179 -j ACCEPT
-          -A INPUT -p tcp --dport 2379 -j ACCEPT
-          -A INPUT -p tcp --dport 2380 -j ACCEPT
-          -A INPUT -p tcp --dport 2381 -j ACCEPT
+          %{ for management_cidr in management_cidrs ~}
+          -A INPUT -s ${management_cidr} -p tcp --dport 22 -j ACCEPT
```
Comment (rata) [inline assets/lokomotive-kubernetes/packet/flatcar-linux/kubernetes/cl/controller.yaml.tmpl:270]: Grr, I'd say we disable ssh (and it will be open when calico starts) but bootkube will fail :-/

File: assets/lokomotive-kubernetes/packet/flatcar-linux/kubernetes/controllers.tf [hunk 2]
```diff
@@ -62,36 +62,21 @@ resource "packet_device" "controllers" {
...
-  content = data.template_file.controller-install[count.index].rendered
-}
-
-data "template_file" "controller-install" {
-  count    = var.controller_count
-  template = file("${path.module}/cl/controller-install.yaml.tmpl")
-
-  vars = {
+  count = var.controller_count
+  content = templatefile("${path.module}/cl/controller-install.yaml.tmpl", {
```
Comment (rata) [inline assets/lokomotive-kubernetes/packet/flatcar-linux/kubernetes/controllers.tf:66]: why is this changed? The linked issue doesn't say anything either. Is it a simplification at the terraform layer with no effect on the generated config?

File: assets/lokomotive-kubernetes/packet/flatcar-linux/kubernetes/calico-host-protection.yaml.tmpl [hunk 3]
```diff
@@ -3,7 +3,7 @@ clusterCIDRs:
 - ${cidr}
 %{ endfor ~}
 managementCIDRs:
-%{ for cidr in management_cidrs ~}
+%{~ for cidr in management_cidrs }
```
Comment (rata) [inline assets/lokomotive-kubernetes/packet/flatcar-linux/kubernetes/calico-host-protection.yaml.tmpl:6]: Was it broken before? How? Or is it a cosmetic change with no effect to the managementCIDRs effectively used? Can you please clarify? :)

File: assets/lokomotive-kubernetes/packet/flatcar-linux/kubernetes/workers/workers.tf [hunk 4]
```diff
@@ -20,22 +20,16 @@ resource "packet_device" "nodes" {
...
-data "ct_config" "install-ignitions" {
-  content = data.template_file.install.rendered
-}
-
 # These configs are used for the fist boot, to run flatcar-install
-data "template_file" "install" {
-  template = file("${path.module}/cl/install.yaml.tmpl")
-
-  vars = {
+data "ct_config" "install-ignitions" {
```
Comment (rata) [inline assets/lokomotive-kubernetes/packet/flatcar-linux/kubernetes/workers/workers.tf:24]: Don't see the need to split both commits in workers and controllers, if they are basically doing the same thing. IMHO, it is a little harder to review just checking if the same was done in the other commit.

File: assets/lokomotive-kubernetes/packet/flatcar-linux/kubernetes/workers/cl/worker.yaml.tmpl [hunk 5]
```diff
@@ -315,10 +315,12 @@ storage:
           :FORWARD ACCEPT [0:0]
           :OUTPUT ACCEPT [0:0]
           -A INPUT -i lo -j ACCEPT
-          -A INPUT -p tcp --dport 22 -j ACCEPT
-          -A INPUT -p tcp --dport 179 -j ACCEPT
-          -A INPUT -p tcp --dport 10250 -j ACCEPT
-          -A INPUT -p tcp --dport 10256 -j ACCEPT
+          %{ for management_cidr in management_cidrs ~}
+          -A INPUT -s ${management_cidr} -p tcp --dport 22 -j ACCEPT
+          %{~ endfor }
```
Comment (rata) [inline assets/lokomotive-kubernetes/packet/flatcar-linux/kubernetes/workers/cl/worker.yaml.tmpl:320]: I don't think ssh is needed for workers before they are running calico (out of band console can be used if needed). What do you think?

File: assets/lokomotive-kubernetes/packet/flatcar-linux/kubernetes/cl/controller.yaml.tmpl [hunk 6]
```diff
@@ -266,11 +266,13 @@ storage:
...
+          %{~ endfor }
+          -A INPUT -s ${node_private_cidr} -p tcp --dport 179 -j ACCEPT
+          -A INPUT -s ${node_private_cidr} -p tcp --dport 2379 -j ACCEPT
+          -A INPUT -s ${node_private_cidr} -p tcp --dport 2380 -j ACCEPT
+          -A INPUT -s ${node_private_cidr} -p tcp --dport 2381 -j ACCEPT
```
Comment (rata) [inline assets/lokomotive-kubernetes/packet/flatcar-linux/kubernetes/cl/controller.yaml.tmpl:275]: Not sure if we should use the `node_private_cidr` or just `10.0.0.0/8`. I like the simplicity of just using `10.0.0.0/8`, but... no strong opinion.

What do you think?

File: assets/lokomotive-kubernetes/packet/flatcar-linux/kubernetes/controllers.tf (hunk 2 above)
Comment (invidian) [reply assets/lokomotive-kubernetes/packet/flatcar-linux/kubernetes/controllers.tf:66]: with tf 0.12, you no longer need `template_file` baked by 3rd party provider, as Terraform provides built-in `templatefile`, so it's recommended ot use this one.

Comment (rata) [reply assets/lokomotive-kubernetes/packet/flatcar-linux/kubernetes/controllers.tf:66]: Oh, okay. I see in [the docs](https://www.terraform.io/docs/providers/template/d/file.html) that it says so. Linking and explaining in the commit would be nice :)

Comment (invidian) [reply assets/lokomotive-kubernetes/packet/flatcar-linux/kubernetes/controllers.tf:66]: I'll add it, though I'm not fan of duplicating this information, as there will be more commits in the future relating to the same problem. We use GitHub, we should make use of it :smile:

File: assets/lokomotive-kubernetes/packet/flatcar-linux/kubernetes/workers/cl/worker.yaml.tmpl (hunk 5 above)
Comment (invidian) [reply assets/lokomotive-kubernetes/packet/flatcar-linux/kubernetes/workers/cl/worker.yaml.tmpl:320]: It seems reasonable, though opinionated, so it should be documented. If calico fails to start for some reason, it makes it more difficult to debug. What do you think @kinvolk/team-cloud?

File: assets/lokomotive-kubernetes/packet/flatcar-linux/kubernetes/workers/workers.tf (hunk 4 above)
Comment (invidian) [reply assets/lokomotive-kubernetes/packet/flatcar-linux/kubernetes/workers/workers.tf:24]: True, will squash.

File: assets/lokomotive-kubernetes/packet/flatcar-linux/kubernetes/cl/controller.yaml.tmpl (hunk 1 above)
Comment (invidian) [reply assets/lokomotive-kubernetes/packet/flatcar-linux/kubernetes/cl/controller.yaml.tmpl:270]: We actually copy secrets over SSH to all controller nodes, not only to the bootstrap one :/

File: assets/lokomotive-kubernetes/packet/flatcar-linux/kubernetes/cl/controller.yaml.tmpl (hunk 6 above)
Comment (invidian) [reply assets/lokomotive-kubernetes/packet/flatcar-linux/kubernetes/cl/controller.yaml.tmpl:275]: `node_private_cidr` could default to `10.0.0.0/8` maybe?

File: assets/lokomotive-kubernetes/packet/flatcar-linux/kubernetes/workers/cl/worker.yaml.tmpl (hunk 5 above)
Comment (invidian) [reply assets/lokomotive-kubernetes/packet/flatcar-linux/kubernetes/workers/cl/worker.yaml.tmpl:320]: Hm, maybe port 22 should also be allowed from `node_private_cidr`?

File: assets/lokomotive-kubernetes/packet/flatcar-linux/kubernetes/controllers.tf (hunk 2 above)
Comment (rata) [reply assets/lokomotive-kubernetes/packet/flatcar-linux/kubernetes/controllers.tf:66]: Don't know what you mean. The linked issue is not clear to me that upstream recommends not using it anymore.

It's easier to review if you have the context you need. If you prefer to link to the issue only, instead of adding some lines, fine (is very easy to make it easier for reviewers by including those lines, but if you really don't want to do it, don't.). But let's make the issue clear on that, at least.

File: assets/lokomotive-kubernetes/packet/flatcar-linux/kubernetes/cl/controller.yaml.tmpl (hunk 6 above)
Comment (rata) [reply assets/lokomotive-kubernetes/packet/flatcar-linux/kubernetes/cl/controller.yaml.tmpl:275]: Now that I think about it, this has been discussed in other issues, I think. The problem is that these rules can't be changed... so, it's complicated. I think both are fine for me (`10.0.0.0/8` vs `node_private_cidr`). Do whatever you consider best :)

File: assets/lokomotive-kubernetes/packet/flatcar-linux/kubernetes/cl/controller.yaml.tmpl (hunk 1 above)
Comment (rata) [reply assets/lokomotive-kubernetes/packet/flatcar-linux/kubernetes/cl/controller.yaml.tmpl:270]: Yeah, exactly :-/

Totally out of scope, but I'd like to play with adding those to ignition compressed (it seems ignition supports compression, not sure if packet will get in the way or if we are using something older, etc.)

Comment (rata) [reply assets/lokomotive-kubernetes/packet/flatcar-linux/kubernetes/cl/controller.yaml.tmpl:270]: But yeah, nothing to do here with this, we need it for now :-/

File: assets/lokomotive-kubernetes/packet/flatcar-linux/kubernetes/workers/workers.tf (hunk 4 above)
Comment (rata) [reply assets/lokomotive-kubernetes/packet/flatcar-linux/kubernetes/workers/workers.tf:24]: Thanks!

File: assets/lokomotive-kubernetes/packet/flatcar-linux/kubernetes/calico-host-protection.yaml.tmpl (hunk 3 above)
Comment (johananl) [reply assets/lokomotive-kubernetes/packet/flatcar-linux/kubernetes/calico-host-protection.yaml.tmpl:6]: In this specific case there is no difference between `{~ }` and `{ ~}` AFAICT. The former deletes the newline on the preceding line and the latter deletes the newline on the same line. It *would* matter if the following line was indented, for example.

File: assets/lokomotive-kubernetes/packet/flatcar-linux/kubernetes/workers/cl/worker.yaml.tmpl (hunk 5 above)
Comment (johananl) [reply assets/lokomotive-kubernetes/packet/flatcar-linux/kubernetes/workers/cl/worker.yaml.tmpl:320]: - IMO we want to be able to SSH into workers at all times, and especially during failures.
- I don't think we want to allow SSH from `node_private_cidr`. This increases the nodes' attack surface (very theoretically) but mainly I don't find it useful. My guess is you're suggesting this to allow using a bastion host (right?), in which case I would include the bastion's `/32` address in `management_cidrs` because it's a management CIDR.

File: assets/lokomotive-kubernetes/packet/flatcar-linux/kubernetes/calico-host-protection.yaml.tmpl (hunk 3 above)
Comment (invidian) [reply assets/lokomotive-kubernetes/packet/flatcar-linux/kubernetes/calico-host-protection.yaml.tmpl:6]: The difference shows up, when there is no `management_cidrs` specified, then:
```yaml
managementCIDRs:


```
is produced, instead of:
```yaml
managementCIDRs:
```

Not a big difference, but I spotted it as a unnecessary new line when testing the template. Can drop this patch as well actually. 

Also added explanation in the commit message.

File: assets/lokomotive-kubernetes/packet/flatcar-linux/kubernetes/controllers.tf (hunk 2 above)
Comment (invidian) [reply assets/lokomotive-kubernetes/packet/flatcar-linux/kubernetes/controllers.tf:66]: Added a bit more context to both issue and commit message :)

File: assets/lokomotive-kubernetes/packet/flatcar-linux/kubernetes/workers/workers.tf (hunk 4 above)
Comment (invidian) [reply assets/lokomotive-kubernetes/packet/flatcar-linux/kubernetes/workers/workers.tf:24]: Done

File: assets/lokomotive-kubernetes/packet/flatcar-linux/kubernetes/cl/controller.yaml.tmpl (hunk 6 above)
Comment (invidian) [reply assets/lokomotive-kubernetes/packet/flatcar-linux/kubernetes/cl/controller.yaml.tmpl]: Decided to use `10.0.0.0/8`. Seems like a sane bootstrap value.

**Comment by invidian** [comment]
Addressed all review feedback, please have a look again.

**Comment by invidian** [comment]
>LGTM. However, I'd add a comment saying why we use 10.0.0.0/8 as that is a trck not obvious for most readers not very involved with Packet networking internals or any new comer.

Make sense, done.

File: assets/lokomotive-kubernetes/packet/flatcar-linux/kubernetes/calico-host-protection.yaml.tmpl (hunk 3 above)
Comment (rata) [reply assets/lokomotive-kubernetes/packet/flatcar-linux/kubernetes/calico-host-protection.yaml.tmpl:6]: Thanks!

File: assets/lokomotive-kubernetes/packet/flatcar-linux/kubernetes/controllers.tf (hunk 2 above)
Comment (rata) [reply assets/lokomotive-kubernetes/packet/flatcar-linux/kubernetes/controllers.tf:70]: Thanks!
